src/index.ts

```diff
@@ -0,0 +1,507 @@
+    if (mac.toString('hex') !== json.crypto.mac) {
+      throw new Error('Key derivation failed - possibly wrong passphrase')
+    }
+
+    const decipher = crypto.createDecipheriv(
+      json.crypto.cipher,
+      derivedKey.slice(0, 16),
+      Buffer.from(json.crypto.cipherparams.iv, 'hex'),
+    )
+    const seed = runCipherBuffer(decipher, ciphertext)
+    return new Wallet(seed)
+  }
+
+  /*
+   * Based on https://github.com/ethereum/pyethsaletool/blob/master/pyethsaletool.py
+   * JSON fields: encseed, ethaddr, btcaddr, email
+   */
+  public static fromEthSale(input: string | EthSaleKeystore, password: string): Wallet {
+    const json: EthSaleKeystore = typeof input === 'object' ? input : JSON.parse(input)
+
+    const encseed = Buffer.from(json.encseed, 'hex')
+
+    // key derivation
+    const derivedKey = crypto.pbkdf2Sync(password, password, 2000, 32, 'sha256').slice(0, 16)
+
+    // seed decoding (IV is first 16 bytes)
+    // NOTE: crypto (derived from openssl) when used with aes-*-cbc will handle PKCS#7 padding internally
+    //       see also http://stackoverflow.com/a/31614770/4964819
+    const decipher = crypto.createDecipheriv('aes-128-cbc', derivedKey, encseed.slice(0, 16))
+    const seed = runCipherBuffer(decipher, encseed.slice(16))
+
+    const wallet = new Wallet(ethUtil.keccak256(seed))
+    if (wallet.getAddress().toString('hex') !== json.ethaddr) {
+      throw new Error('Decoded key mismatch - possibly wrong passphrase')
+    }
+    return wallet
+  }
+
+  // private getters
+
+  private get pubKey(): Buffer {
+    if (!keyExists(this.publicKey)) {
+      this.publicKey = ethUtil.privateToPublic(this.privateKey as Buffer)
+    }
+    return this.publicKey
+  }
+
+  private get privKey(): Buffer {
+    if (!keyExists(this.privateKey)) {
+      throw new Error('This is a public key only wallet')
+    }
+    return this.privateKey
+  }
+
+  // public instance methods
+
+  public getPrivateKey(): Buffer {
+    return this.privKey
+  }
+
+  public getPrivateKeyString(): string {
+    return ethUtil.bufferToHex(this.privKey)
+  }
+
+  public getPublicKey(): Buffer {
+    return this.pubKey
+  }
+
+  public getPublicKeyString(): string {
+    return ethUtil.bufferToHex(this.getPublicKey())
+  }
+
+  public getAddress(): Buffer {
+    return ethUtil.publicToAddress(this.pubKey)
+  }
+
+  public getAddressString(): string {
+    return ethUtil.bufferToHex(this.getAddress())
+  }
+
+  public getChecksumAddressString(): string {
+    return ethUtil.toChecksumAddress(this.getAddressString())
+  }
+
+  public toV3(password: string, opts?: Partial<V3Params>): V3Keystore {
+    if (!keyExists(this.privateKey)) {
+      throw new Error('This is a public key only wallet')
+    }
+
+    const params = mergeToV3ParamsWithDefaults(opts)
+    const kdfParams = mergeKDFParamsWithDefaults(params)
+
+    let derivedKey: Buffer
+    let finalKDFParams: KDFParams
+
+    if (params.kdf === 'pbkdf2') {
```

Comment:
I feel like this kdfParams processing has gotten more complex than necessary (with 2 steps, first merging from v3Params with defaults, and later stripping unused fields). One option I see is to ditch `mergeKDFParamsWithDefaults` and `kdfParamsForX`, and have a one step e.g. `scryptFromV3Params`. Or even simply instantiate a `XKDFParams` in each of the kdf branches.

Reply:
I reckon you're right. I was trying very hard to avoid a flow refactor, but the result is ugliness. Will go with option B: an `XKDFParams` for each function type